src/protocol/clock/extent.rs

```diff
@@ -0,0 +1,113 @@
+use std::num::{IntErrorKind, TryFromIntError};
+use std::time::Duration;
+
+/// Count fixed sized increments. Provides the product of the increment and its count.
+pub trait IncrementCounter: Sized + Default {
+    type Increment;
+
+    type IncrementsCount;
+    type IncrementIncrementsProduct;
+
+    fn new(size: &Self::Increment, quantity: &Self::IncrementsCount) -> Self;
+
+    fn add(&self, add: Self::IncrementsCount) -> Result<Self, IntErrorKind>;
+    fn sub(&self, sub: Self::IncrementsCount) -> Result<Self, IntErrorKind>;
+
+    fn product_current(&self) -> Result<Option<Self::IncrementIncrementsProduct>, TryFromIntError>;
+    fn product_next(&self) -> Result<Option<Self::IncrementIncrementsProduct>, TryFromIntError>;
+}
+
+pub type ExtentIncrement = Duration;
+pub type ExtentIncrementSeconds = u64;
+pub type ExtentIncrementsCount = u64;
+pub type ExtentIncrementIncrementProduct = ExtentIncrement;
+
+#[derive(Debug, Default, Hash, PartialEq, Eq)]
+pub struct Extent {
+    pub size: ExtentIncrement,
+    pub quantity: ExtentIncrementsCount,
+}
+
+impl Extent {
+    pub const fn from_sec(size: ExtentIncrementSeconds, quantity: &ExtentIncrementsCount) -> Self {
+        Self {
+            size: ExtentIncrement::from_secs(size),
+            quantity: *quantity,
+        }
+    }
+}
+
+impl IncrementCounter for Extent {
+    type Increment = ExtentIncrement;
+    type IncrementsCount = ExtentIncrementsCount;
+    type IncrementIncrementsProduct = ExtentIncrementIncrementProduct;
+
+    fn new(size: &Self::Increment, quantity: &Self::IncrementsCount) -> Self {
+        Self {
+            size: *size,
+            quantity: *quantity,
+        }
+    }
+
+    fn add(&self, add: Self::IncrementsCount) -> Result<Self, IntErrorKind> {
+        match self.quantity.checked_add(add) {
+            None => Err(IntErrorKind::PosOverflow),
+            Some(quantity) => Ok(Self {
+                size: self.size,
+                quantity,
+            }),
+        }
+    }
```

Comment:
@da2ce7 I think I would call this method `extend` or `expand` because with `add` it seems you are adding to structs of the same type.

For the `sub` it could be `collapse` or `shorten`.

Reply:
Hmm... This is something that I need to consider.

Reply:
I've decided on `increase` and `decrease`.